library/core/src/clone.rs

```diff
@@ -124,9 +124,70 @@ pub trait Clone: Sized {
 
     /// Performs copy-assignment from `source`.
     ///
-    /// `a.clone_from(&b)` is equivalent to `a = b.clone()` in functionality,
+    /// `a.clone_from(&b)` is equivalent to `a = b.clone()` in functionality except panic safety,
     /// but can be overridden to reuse the resources of `a` to avoid unnecessary
     /// allocations.
+    ///
+    /// # Panic safety
+    ///
+    /// Due to it's nature, this method is not panic safe.
+    /// If call panics, `self` can be left in inconsistent state.
+    /// This is different from `a = b.clone()` because in that case `a` is overwritten
+    /// only if `clone()` succeedes.
+    ///
+    /// `clone_from` is intended to preserve resources owned by `self`
+    /// so it cannot guarantee panic safety.
+    ///
+    /// In that example, second assert ***can*** fail:
+    /// ```no_run
+    /// use std::sync::atomic::{AtomicBool, Ordering};
+    ///
+    /// // Ensure that static vector accessed only once
+    /// static WAS_CALLED: AtomicBool = AtomicBool::new(false);
+    /// if WAS_CALLED.fetch_or(true, Ordering::Acquire){
+    ///     return;
+    /// }
+    ///
+    /// #[derive(PartialEq, Eq, Debug)]
+    /// struct IAmPanicDuringClone;
+    ///
+    /// impl Clone for IAmPanicDuringClone {
+    ///     fn clone(&self) -> Self {
+    ///         panic!()
+    ///     }
+    /// }
+    ///
+    /// // Here little complicated
+    /// // because we need to pass it through panic boundary for demo purposes.
+    /// // Hacky and improper way to do this is usage of static variable.
+    /// // Don't write like that in production.
+    /// static mut A: Vec<IAmPanicDuringClone> = Vec::new();
+    /// let a = unsafe{
+    ///     // Safe because we have AtomicBool check before.
+    ///     &mut A
+    /// };
+    ///
+    /// *a = vec![
+    ///         IAmPanicDuringClone,
+    ///         IAmPanicDuringClone,
+    ///         IAmPanicDuringClone,
+    ///     ];
+    /// let b = vec![
+    ///     IAmPanicDuringClone,
+    ///     IAmPanicDuringClone,
+    ///     IAmPanicDuringClone,
+    /// ];
+    /// let c = vec![IAmPanicDuringClone, IAmPanicDuringClone];
+    ///
+    /// // This succeedes.
+    /// assert_eq!(a, &b);
+    /// std::panic::catch_unwind(|| {
+    ///     unsafe { A.clone_from(&c) };
+    /// }).expect_err("Must panic");
+    ///
+    /// // This fails at Rust 1.61.
+    /// assert_eq!(unsafe { &A }, &b);
+    /// ```
```

Comment:
I'm not comfortable with using this as an example (it feels a bit contrived), how do you feel about using this version instead?

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=85975c6d68d63fd14d816b7849d10c56

Reply:
Absolutely agree, your version is much more cleaner and has less noise in it.

Would fix PR later.